engine-standalone-tracing/src/types/call_tracer.rs

```diff
@@ -0,0 +1,288 @@
+//! This module defines data structure to produce traces compatible with geths "callTracer":
+//! https://github.com/ethereum/go-ethereum/blob/ad15050c7fbedd0f05a49e81400de18c2cc2c284/eth/tracers/native/call.go
+
+use aurora_engine_types::{types::Address, U256};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CallFrame {
+    call_type: CallType,
+    from: Address,
+    to: Option<Address>,
+    value: U256,
+    gas: u64,
+    gas_used: u64,
+    input: Vec<u8>,
+    output: Vec<u8>,
+    error: Option<String>,
+    calls: Vec<CallFrame>,
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct CallTracer {
+    call_stack: Vec<CallFrame>,
+}
+
+impl CallTracer {
+    fn end(&mut self, output: Vec<u8>, error: Option<&evm::ExitReason>) {
+        let frame = self.call_stack.first_mut().unwrap();
+        match error {
+            None => frame.output = output,
+            Some(error) => {
+                let error_message = format!("{:?}", error);
+                if error_message.to_lowercase().contains("revert") {
+                    frame.output = output;
+                }
+                frame.error = Some(error_message);
+            }
+        }
+    }
+
+    fn enter(
+        &mut self,
+        call_type: CallType,
+        from: Address,
+        to: Address,
+        input: Vec<u8>,
+        gas: u64,
+        value: U256,
+    ) {
```

Comment:
I don't suppose we need to skip tracing if it was interrupted here?

Reply:
I think all that interrupt handling is needed because tracing can be done async in the Go code, so they need a way to propagate cancellation from other async tasks. For us, all our tracing code is synchronous.